src/core/lib/transport/metadata.h

```diff
@@ -241,42 +268,34 @@ class AllocatedMetadata {
   AllocatedMetadata(const grpc_slice& key, const grpc_slice& value);
   ~AllocatedMetadata();
 
-  const grpc_slice& key() const { return key_; }
-  const grpc_slice& value() const { return value_; }
+  const grpc_slice& key() const { return common_.key; }
+  const grpc_slice& value() const { return common_.value; }
   UserData* user_data() { return &user_data_; }
 
 #ifndef NDEBUG
   void Ref(const char* file, int line) {
-    grpc_mdelem_trace_ref(this, key_, value_, RefValue(), file, line);
-    Ref();
+    grpc_mdelem_trace_ref(this, common_.key, common_.value, RefValue(), file,
+                          line);
+    common_.Ref();
   }
   bool Unref(const char* file, int line) {
-    grpc_mdelem_trace_unref(this, key_, value_, RefValue(), file, line);
-    return Unref();
+    grpc_mdelem_trace_unref(this, common_.key, common_.value, RefValue(), file,
+                            line);
+    return common_.Unref();
   }
 #endif  // ifndef NDEBUG
-  void Ref() {
-    /* we can assume the ref count is >= 1 as the application is calling
-       this function - meaning that no adjustment to mdtab_free is necessary,
-       simplifying the logic here to be just an atomic increment */
-    refcnt_.FetchAdd(1, MemoryOrder::RELAXED);
-  }
-  bool Unref() {
-    const intptr_t prior = refcnt_.FetchSub(1, MemoryOrder::ACQ_REL);
-    GPR_DEBUG_ASSERT(prior > 0);
-    return prior == 1;
+  void Ref() { common_.Ref(); }
+  bool Unref() { return common_.Unref(); }
+
+  static constexpr size_t offset() {
+    return offsetof(AllocatedMetadata, common_);
   }
 
  private:
-  intptr_t RefValue() { return refcnt_.Load(MemoryOrder::RELAXED); }
+  intptr_t RefValue() { return common_.RefValue(); }
 
   /* must be byte compatible with grpc_mdelem_data */
-  grpc_slice key_;
-  grpc_slice value_;
-
-  /* private only data */
-  grpc_core::Atomic<intptr_t> refcnt_;
-
+  MdCommon common_;
```

Comment:
This is going to be fragile. I would suggest renaming this to `MdBase` inheriting from it so that we are sure all meta-data types have the same first fields. As a bonus you won't need to repeat methods like Ref()/Unref()/RefValue() everywhere.

Reply:
Renamed to RefcountedMdBase.